ta/pkcs11/src/object.c

```diff
@@ -409,3 +409,591 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+	if (!attr) {
+		rc = PKCS11_CKR_DEVICE_MEMORY;
+		goto out;
+	}
+
+	res = TEE_ReadObjectData(hdl, attr, info.dataSize, &read_bytes);
+	if (!res) {
+		res = TEE_SeekObjectData(hdl, 0, TEE_DATA_SEEK_SET);
+		if (res)
+			EMSG("Seek to 0 failed with %#"PRIx32, res);
+	}
+
+	if (res) {
+		rc = tee2pkcs_error(res);
+		EMSG("Read %"PRIu32" bytes, failed %#"PRIx32,
+		     read_bytes, res);
+		goto out;
+	}
+	if (read_bytes != info.dataSize) {
+		EMSG("Read %"PRIu32" bytes, expected %"PRIu32,
+		     read_bytes, info.dataSize);
+		rc = PKCS11_CKR_GENERAL_ERROR;
+		goto out;
+	}
+
+	if (!attributes_match_reference(attr, req_attrs)) {
+		rc = PKCS11_RV_NOT_FOUND;
+		goto out;
+	}
+
+	obj->attributes = attr;
+	attr = NULL;
+	obj->attribs_hdl = hdl;
+	hdl = TEE_HANDLE_NULL;
+
+	rc = PKCS11_CKR_OK;
+
+out:
+	TEE_Free(attr);
+	if (obj->attribs_hdl == TEE_HANDLE_NULL && hdl != TEE_HANDLE_NULL)
+		TEE_CloseObject(hdl);
+
+	return rc;
+}
+
+static void release_find_obj_context(struct pkcs11_session *session,
+				     struct pkcs11_find_objects *find_ctx)
+{
+	size_t idx = 0;
+
+	if (!find_ctx)
+		return;
+
+	/* Release handles not yet published to client */
+	idx = find_ctx->next;
+	if (idx < find_ctx->temp_start)
+		idx = find_ctx->temp_start;
+
+	for (; idx < find_ctx->count; idx++)
+		handle_put(&session->object_handle_db, find_ctx->handles[idx]);
+
+	TEE_Free(find_ctx->attributes);
+	TEE_Free(find_ctx->handles);
+	TEE_Free(find_ctx);
+}
+
+enum pkcs11_rc entry_find_objects_init(struct pkcs11_client *client,
+				       uint32_t ptypes, TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *ctrl = params;
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	struct serialargs ctrlargs = { };
+	struct pkcs11_session *session = NULL;
+	struct pkcs11_object_head *template = NULL;
+	struct obj_attrs *req_attrs = NULL;
+	struct pkcs11_object *obj = NULL;
+	struct pkcs11_find_objects *find_ctx = NULL;
+
+	if (!client || ptypes != exp_pt)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
+
+	rc = serialargs_get_session_from_handle(&ctrlargs, client, &session);
+	if (rc)
+		return rc;
+
+	rc = serialargs_alloc_get_attributes(&ctrlargs, &template);
+	if (rc)
+		return rc;
+
+	if (serialargs_remaining_bytes(&ctrlargs)) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	/* Search objects only if no operation is on-going */
+	if (session_is_active(session)) {
+		rc = PKCS11_CKR_OPERATION_ACTIVE;
+		goto out;
+	}
+
+	if (session->find_ctx) {
+		EMSG("Active object search already in progress");
+		rc = PKCS11_CKR_FUNCTION_FAILED;
+		goto out;
+	}
+
+	/* Must zero init the structure */
+	find_ctx = TEE_Malloc(sizeof(*find_ctx), TEE_MALLOC_FILL_ZERO);
+	if (!find_ctx) {
+		rc = PKCS11_CKR_DEVICE_MEMORY;
+		goto out;
+	}
+
+	rc = sanitize_client_object(&req_attrs, template,
```

Comment:
Should we first sanitize input and after that allocate find_ctx? Eg. switch order of `TEE_Malloc()` and `sanitize_client_object()`.

Reply:
Yes, I will do that

Reply:
smart, thanks.